Sources/_StringProcessing/Engine/MEBuiltins.swift

```diff
@@ -1,13 +1,221 @@
 
 extension Processor {
+  mutating func matchBuiltin(
+    _ cc: _CharacterClassModel.Representation,
+    _ isInverted: Bool,
+    _ isStrictAscii: Bool
+  ) -> Bool {
+    guard let c = load() else {
+      signalFailure()
+      return false
+    }
+
+    var matched: Bool
+    var next = input.index(after: currentPosition)
+    switch cc {
+    case .any, .anyGrapheme: matched = true
+    case .anyScalar:
+      matched = true
+      next = input.unicodeScalars.index(after: currentPosition)
+    case .digit:
+      matched = c.isNumber && (c.isASCII || !isStrictAscii)
+    case .horizontalWhitespace:
+      matched = c.unicodeScalars.first?.isHorizontalWhitespace == true
+      && (c.isASCII || !isStrictAscii)
+    case .newlineSequence, .verticalWhitespace:
+      matched = c.unicodeScalars.first?.isNewline == true
+      && (c.isASCII || !isStrictAscii)
+    case .whitespace:
+      matched = c.isWhitespace && (c.isASCII || !isStrictAscii)
+    case .word:
+      matched = c.isWordCharacter && (c.isASCII || !isStrictAscii)
+    }
+    if isInverted {
+      matched.toggle()
+    }
+    if matched {
+      currentPosition = next
+      return true
+    } else {
+      signalFailure()
+      return false
+    }
+  }
+  
+  mutating func matchBuiltinScalar(
+    _ cc: _CharacterClassModel.Representation,
+    _ isInverted: Bool,
+    _ isStrictAscii: Bool
+  ) -> Bool {
+    guard let c = loadScalar() else {
+      signalFailure()
+      return false
+    }
+
+    var matched: Bool
+    var next = input.unicodeScalars.index(after: currentPosition)
+    switch cc {
+    case .any: matched = true
+    case .anyScalar: matched = true
+    case .anyGrapheme:
+      matched = true
+      next = input.index(after: currentPosition)
+    case .digit:
+      matched = c.properties.numericType != nil && (c.isASCII || !isStrictAscii)
+    case .horizontalWhitespace:
+      matched = c.isHorizontalWhitespace && (c.isASCII || !isStrictAscii)
+    case .verticalWhitespace:
+      matched = c.isNewline && (c.isASCII || !isStrictAscii)
+    case .newlineSequence:
+      matched = c.isNewline && (c.isASCII || !isStrictAscii)
+      if c == "\r" && next != input.endIndex && input.unicodeScalars[next] == "\n" {
+        input.unicodeScalars.formIndex(after: &next)
+      }
+    case .whitespace:
+      matched = c.properties.isWhitespace && (c.isASCII || !isStrictAscii)
+    case .word:
+      matched = (c.properties.isAlphabetic || c == "_") && (c.isASCII || !isStrictAscii)
+    }
+    if isInverted {
+      matched.toggle()
+    }
+    if matched {
+      currentPosition = next
+      return true
+    } else {
+      signalFailure()
+      return false
+    }
+  }
+  
+  func isAtStartOfLine(_ payload: AssertionPayload) -> Bool {
+    if currentPosition == subjectBounds.lowerBound { return true }
+    switch payload.semanticLevel {
+    case .graphemeCluster:
+      return input[input.index(before: currentPosition)].isNewline
```

Comment:
CC @Azoy do we have a test case for when this starts with a non-grapheme-cluster aligned lower bound?

Reply:
I'm unsure if we do or not. How would we go about adding a test for this?

Reply:
On a related note: what should we do about the `.anyScalar` class in grapheme mode? If we forbid that in grapheme mode then we would always be grapheme-cluster aligned unless a user custom consumer returns a non-aligned index, and we could always forbid that as well

Reply:
As for the test I don't think we could ever have a non-grapheme aligned `subjectBounds` or `searchBounds` since we always handle creating and stepping the bounds

```
    var low = searchBounds.lowerBound
    let high = searchBounds.upperBound
    while true {
      if let m: Regex<Output>.Match = try _match(
        input, from: low, using: &cpu
      ) {
        return m
      }
      if low >= high { return nil }
      if graphemeSemantic {
        input.formIndex(after: &low)
      } else {
        input.unicodeScalars.formIndex(after: &low)
      }
      cpu.reset(currentPosition: low)
    }
```

Reply:
That's right, there should be a grapheme boundary assertion whenever stepping up into grapheme semantic mode.